gemrb/core/GUI/EventMgr.cpp

```diff
@@ -19,501 +19,479 @@
+			ke.keycode = GEM_LEFT;
+			break;
+		case CONTROLLER_BUTTON_DPAD_RIGHT:
+			ke.keycode = GEM_RIGHT;
 			break;
-		}
-	}
-	if (topwin.size() != 0) {
-			topwin.insert( topwin.begin(), Index );
-	} else {
-		topwin.push_back( Index );
 	}
+	
+	return ke;
 }
 
-void EventMgr::SetDefaultFocus(Window *win)
+bool EventMgr::ModState(unsigned short mod)
 {
-	if (!last_win_focused) {
-		last_win_focused = win;
-		last_win_focused->SetFocused(last_win_focused->GetControl(0));
-	}
-	last_win_over = NULL;
+	return (modKeys & buttonbits(mod)).any();
 }
 
-/** Adds a Window to the Event Manager */
-void EventMgr::AddWindow(Window* win)
+bool EventMgr::MouseButtonState(EventButton btn)
 {
-	unsigned int i;
+	return (mouseButtonFlags & buttonbits(btn)).any();
+}
 
-	if (win == NULL) {
-		return;
-	}
-	bool found = false;
-	for (i = 0; i < windows.size(); i++) {
-		if (windows[i] == win) {
-			goto ok;
-		}
-		if(windows[i]==NULL) {
-			windows[i] = win;
-ok:
-			SetOnTop( i );
-			found = true;
-			break;
-		}
-	}
-	if (!found) {
-		windows.push_back( win );
-		if (windows.size() == 1)
-			topwin.push_back( 0 );
-		else
-			SetOnTop( ( int ) windows.size() - 1 );
-	}
-	SetDefaultFocus(win);
+bool EventMgr::MouseDown()
+{
+	return mouseButtonFlags.any();
 }
-/** Frees and Removes all the Windows in the Array */
-void EventMgr::Clear()
+
+bool EventMgr::FingerDown()
 {
-	topwin.clear();
-	windows.clear();
-	last_win_focused = NULL;
-	last_win_mousefocused = NULL;
-	last_win_over = NULL;
-	function_bar = NULL;
+	return fingerStates.size() > 0;
 }
 
-/** Remove a Window from the array */
-void EventMgr::DelWindow(Window *win)
+bool EventMgr::ControllerButtonState(EventButton btn)
 {
-	bool focused = (last_win_focused == win);
-	if (focused) {
-		last_win_focused = NULL;
-	}
-	if (last_win_mousefocused == win) {
-		last_win_mousefocused = NULL;
-	}
-	if (last_win_over == win) {
-		last_win_over = NULL;
-	}
-	if (function_bar == win) {
-		function_bar = NULL;
-	}
+	return (controllerButtonStates & buttonbits(btn)).any();
+}
 
-	if (windows.size() == 0) {
+void EventMgr::DispatchEvent(Event& e)
+{
+	if (TouchInputEnabled == false && e.EventMaskFromType(e.type) & Event::AllTouchMask) {
 		return;
 	}
 
-	int pos = -1;
-	std::vector< Window*>::iterator m;
-	for (m = windows.begin(); m != windows.end(); ++m) {
-		pos++;
-		if ( (*m) == win) {
-			(*m) = NULL;
-			std::vector< int>::iterator t;
-			for (t = topwin.begin(); t != topwin.end(); ++t) {
-				if ( (*t) == pos) {
-					topwin.erase( t );
-					if (focused && topwin.size() > 0) {
-						//revert focus to new top window
-						SetFocused(windows[topwin[0]], NULL);
-					}
-					return;
-				}
+	Video* video = core->GetVideoDriver();
+
+	e.time = GetTickCount();
+
+	if (e.type == Event::TextInput) {
+		if (e.text.text.length() == 0) {
+			return;
+		}
+	} else if (e.EventMaskFromType(e.type) & Event::AllKeyMask) {
+		// first check for hot key listeners
+		static unsigned long lastKeyDown = 0;
```

Comment:
Codacy found an issue: [The scope of the variable 'lastKeyDown' can be reduced.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=6310305)